### ISSUE TITLE
Data fetching utilites

### ISSUE THREAD

**Issue by jakubszalek** [adamborowski/react-arch-demo#7]
I noticed that you've chosen to develop your own data fetching utilities - that's nice but in current shape it causes some problems.

1. Two crucial parts of the functionality are split: data fetching and request cancelling - I see that you make use of code composition but it might be problematic for other developers to understand how to use it.
2. `createGraphqlSearchClient` has hardcoded query inside and cannot be reused with other queries. I see that you left a TODO flag there but it's crucial to prepare basic app utilities to be ready for next development steps.
3. Generally data fetching utilities are often used across the whole app so it should be as simple and reusable as possible. I think that current implementation could be simplified. Currently, it seems like every page needs to utilise separate graphQl client with it's own data adapter which makes massive overhead. We should be able to create one global gql client and expose lightweight method that allows us to pass query (and adapter) and fetch the data.

Please consider some refactor in current implementation.

**Comment by adamborowski** [comment]
@mrcodegit 

Re: 1
I am not sure about this point. This API is built on the top of promises which as we know - are not cancelable.
We can make a wrapper that will make canceling in the same place and the API could be like:
```ts
const fetchingPromise = client.search(query)

// fetchingPromise.cancel()
// fetchingPromise.then()
// ...
```

But I realized that this API doesn't really work nice with async/await, on both: consumption and impementation levels.
And let's look at react-query, their `getData()` pattern also requires signal injecting.

Now, after this short context explanation, what would you say about point 1?

**Comment by adamborowski** [comment]
@mrcodegit 
re: 3

I am not sure that the suggestion to move awareness of endpoints data type incompatibility with client types into page components is a good idea - they will become more complex. 
That's the beauty of Page components, they only require the client and are free from IO complexity

**Comment by adamborowski** [comment]
@mrcodegit point 2 addressed in #11
